src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1029 @@
+# Integrate with the kubernetes events API. 
+# This module sends events to Kubernetes, and also captures/tracks all events
+# in the rook-ceph namespace so kubernetes activity like pod restarts,
+# imagepulls etc can be seen from within the ceph cluster itself.
+#
+# To interact with the events API, the mgr service to access needs to be 
+# granted additional permissions
+# e.g. kubectl -n rook-ceph edit clusterrole rook-ceph-mgr-cluster-rules
+#
+# These are the changes needed;
+# - apiGroups:
+#   - ""
+#   resources:
+#   - events
+#   verbs:
+#   - create
+#   - patch
+#   - list
+#   - get
+#   - watch
+
+# LogEntry class dynamically builds an object from a set of args,
+# disable the linter for the no-member error
+# pylint: disable=no-member
+
+
+import os
+import sys
+import time
+import json
+import threading
+
+from datetime import datetime, timedelta, tzinfo
+from urllib3.exceptions import MaxRetryError
+from collections import OrderedDict
+
+import rados
+from mgr_module import MgrModule
+from ceph_argparse import json_command, run_in_thread
+
+try:
+    import queue
+except ImportError:
+    # python 2.7.5
+    import Queue as queue
+finally:
+    # python 2.7.15 or python3
+    event_queue = queue.Queue()
+
+try:
+    from kubernetes import client, config, watch
```

Comment:
This is still unresolved. By just merging this, we're introducing a kubernetes dependency to the ceph-mgr package.

Reply:
Are we? Won't can_run be set to false so the module couldn't be enabled? Like mgr/rook it only comes into play under kubernetes, which means the daemon is running as a container, and the kubernetes client library would have to be included in the container. For example, we have a influx mgr module which requires influxdb package ... isn't this the same thing?

Reply:
@tchaikov What do you think?

Reply:
>  By just merging this, we're introducing a kubernetes dependency to the ceph-mgr package.

i'd say it would be python-kubernetes or python3-kubernetes.

>  For example, we have a influx mgr module which requires influxdb package ... isn't this the same thing?

i'd say it is python-influxdb or python3-influxdb.

hi @pcuzner @sebastian-philipp , i'd like to concur Sebastian's opinion. we see this feature in two ways here:

1. the user is on his/her own, as this is an opt-in choice. so, if he/she wants to use this feature, he/she needs to take care of the dependencies. but i think, we, as both the upstream and maintainer of Ceph, need to help our user to prepare the dependency in a standardized way. when it comes to our building system, i.e., cmake , we need to warn our user if any of dependencies is missing at configure-time. instead of at build time or link time. the same applies to rpm/deb packaging, so ideally, we should ensure that our package is properly packages in the sense that all the runtime dependencies are defined. otherwise, our users will have to check the mgr's logging message for the clues of malfunctioning mgr modules. IMO, i think our users deserve something better.
2. we should prepare packages in ways that are standarized, as it's an opt-in choice. i think that's another reason to package ceph-mgr modules in separated packages. so we will have the chance to define the dependencies for those modules without forcing users to install them even if some of the modules are not used at all. in this case, i think we need to package k8sevents as a separated package just ceph-mgr-rook. so it will depend on python-kubernetes or python3-kubernetes as ceph-mgr-rook does. see https://cbs.centos.org/koji/rpminfo?rpmID=166231 .

Reply:
@tchaikov so you're saying that the PR must include updates to the packaging build process? Would be helpful, if you could provide some guidance on which files need to change (rpm/deb/other?)

Reply:
@pcuzner i don't think the missing packaging is a blocker of this change. we can always do this in a follow-up PR. please 

- search for "mgr-rook" in `ceph.spec.in`
- search for "ceph-mgr-rook" in `debian/control`, and check `debian/ceph-mgr-rook.*` 

and follow the model of it. and you can push the change to ceph-ci to verify if the packaging works. lemme know if anything i can help with. =)

Reply:
thanks @tchaikov